app/models/devise_token_auth/concerns/user.rb

```diff
@@ -257,4 +255,10 @@ def remove_tokens_after_password_reset
     end
   end
 
+  def clean_old_tokens
```

Comment:
Isn't this code technically <code>O(n<sup>2</sup>)</code>? (The `Enumberable#min_by` call is inside a `while` loop).

Even though `Enumberable#min_by` returns a single key/value pair, it still iterates over the entire collection every time it's called.

I would think that it is more efficient to iterate over the collection once, sorting the key/value pairs by `:expiry` of the value, and then shift the oldest key/values from tokens in a while loop.

```ruby
  ...
    protected
    ...
    def max_client_tokens_exceeded?
      tokens.length > DeviseTokenAuth.max_number_of_devices
    end

    def clean_old_tokens
      if tokens.present? && max_client_tokens_exceeded?
        # Using Enumerable#sort_by on a Hash will typecast it into an associative
        #   Array (i.e. an Array of key-value Array pairs). However, since Hashes
        #   have an internal order in Ruby 1.9+, the resulting sorted associative
        #   Array can be converted back into a Hash, while maintaining the sorted
        #   order.
        self.tokens = tokens.sort_by { |_cid, v| v[:expiry] || v['expiry'] }.to_h

        # Since the tokens are sorted by expiry, shift the oldest client token
        #   off the Hash until it no longer exceeds the maximum number of clients
        tokens.shift while max_client_tokens_exceeded?
      end
    end
  ...
```

Reply:
This change breaks randomically the test that you mentioned: `test/controllers/demo_user_controller_test.rb#L506`. I tried for a day fixing it but it's quite weird (and hard to debug with minitest). If we merge the PR without this change, can you create a new one and try to fix it with your comment?

Reply:
Yeah, no problem. I can submit a second PR. Although, I think I saw random test failures on travis-ci before I even made my comment?

Reply:
Yes, but they weren't random, they were my errors, I haven't updated the PR yet

Reply:
Is the debugging a challenge because the `Minitest::Reporters::ProgressReporter` leaves out line numbers? I've definitely been challenged with debugging failing tests while using that reporter.  While I prefer the `ProgressReporter`, I find that the default minitest reporter provides much more useful debugging info.

I usually just temporarily comment out [line 27 of the `test_helper.rb`](https://github.com/lynndylanhurley/devise_token_auth/blob/master/test/test_helper.rb#L27) while running these tests. YMMV.

I've been meaning to look into the the issue and submit a PR to [`kern/minitest-reporters`](https://github.com/kern/minitest-reporters) with a fix.

Reply:
Thanks for the tip! It doesn't throw a backtrace either 😞